File: packages/ui/db/entity/Crag.ts
import { cascadeOneToMany } from "@/db/cascadeOptions";
import { Area } from "@/db/entity/Area";
import BaseColumnSchemaPart from "@/db/entity/BaseColumnSchemaPart";
import BoundsSchema, { Bounds } from "@/db/entity/Bounds";
import { Commentable } from "@/db/entity/Commentable";
import { Photoable } from "@/db/entity/Photoable";
import { type ICrag } from "models";
import { EntitySchema } from "typeorm";
import CoordinateSchema, { Coordinate } from "./Coordinate";

export type Crag = Omit<ICrag, "bounds"> & {
  areas?: Area[];
  center: Coordinate;
  bounds?: Bounds;
  commentable?: Commentable;
  photoable?: Photoable;
};

const CragSchema = new EntitySchema<Crag>({
  name: "crag",
  columns: {
    ...BaseColumnSchemaPart,
    name: {
      type: String,
    },
    description: {
      type: String,
      nullable: true,
    },
    defaultZoom: {
      type: "int",
    },
    minZoom: {
      type: "int",
      nullable: true,
    },
    maxZoom: {
      type: "int",
      nullable: true,
    },
  },
  relations: {
    areas: {
      type: "one-to-many",
      inverseSide: "crag",
      target: "area",
      ...cascadeOneToMany,
    },
    commentable: {
      type: "one-to-one",
      nullable: true,
      target: "commentable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
    photoable: {
      type: "one-to-one",
      nullable: true,
      target: "photoable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
    trail: {
      type: "one-to-one",
      target: "trail",
      joinColumn: true,
      onDelete: "SET NULL",
      cascade: ["insert", "update"],
    },
  },
  embeddeds: {
    center: {
      schema: CoordinateSchema,
      prefix: "center_",
    },
    bounds: {
      schema: BoundsSchema,
      prefix: "bounds_",
    },
  },
});

export default CragSchema;

File: go-server/internal/models/crag.go
package models

type Crag struct {
	ID          int64        `json:"id,omitzero" sqlp:"id"`
	Name        string       `json:"name" sqlp:"name"`
	Description *string      `json:"description,omitzero" sqlp:"description"`
	Bounds      Bounds       `json:"bounds,omitzero" sqlp:"bounds,promote"`
	Center      Coordinate   `json:"center" sqlp:"center,promote"`
	DefaultZoom int          `json:"defaultZoom" sqlp:"defaultZoom"`
	MinZoom     *int         `json:"minZoom,omitzero" sqlp:"minZoom"`
	MaxZoom     *int         `json:"maxZoom,omitzero" sqlp:"maxZoom"`
	Parking     *Parking     `json:"parking,omitzero" sqlp:"parking"`
	Areas       []Area       `json:"areas,omitzero"`
	Commentable *Commentable `json:"commentable,omitzero" sqlp:"commentable"`
	Photoable   *Photoable   `json:"photoable,omitzero" sqlp:"photoable"`
	Trail       *Trail       `json:"trail,omitzero" sqlp:"trail"`
	TrailID     *int64       `json:"-" sqlp:"trailId"`
	Timestamps
}

func (c *Crag) IsZero() bool {
	return (c.ID == 0 &&
		c.Name == "" &&
		c.Description == nil &&
		c.Bounds.IsZero() &&
		c.Center.IsZero() &&
		c.DefaultZoom == 0 &&
		c.MinZoom == nil &&
		c.MaxZoom == nil &&
		c.Parking == nil &&
		len(c.Areas) == 0 &&
		c.Commentable == nil &&
		c.Photoable == nil &&
		c.Trail == nil)
}
